# Support Vector Machines

- This is a supplement material for my lectures on ML. It sheds light on Python implementations of the supervised machine learning algorithms. 
- I assume you know Python syntax and how it works. If you don't, I highly recommend you to take a break and get introduced to the language before going forward with my code. 
- This material is represented as a Jupyter notebook, which you can easily download to reproduce the code and play around with it. 

## 1. Libraries

To build a model, we need 
- `pandas` library to work with panel dataframes
- `sklearn.model_selection.train_test_split` to split the dataset into train and test
- `sklearn.svm.SVC` to build a model

In addition, I'd like to show you `sklearn.preprocessing.MinMaxScaler` - a library for data normalization.

In [1]:
import pandas as pd
from sklearn.svm import SVC
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

## 2. Data Load & Overview

Let's load the dataset and understand it.

In [3]:
# Load the dataset
df = pd.read_csv('https://raw.githubusercontent.com/5x12/ml-cookbook/master/supplements/data/heart.csv')

In [4]:
# Print the first 10 rows of the dataset
df.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   
5   58    0   0       100   248    0        0      122      0      1.0      1   
6   58    1   0       114   318    0        2      140      0      4.4      0   
7   55    1   0       160   289    0        0      145      1      0.8      1   
8   46    1   0       120   249    0        0      144      0      0.8      2   
9   54    1   0       122   286    0        0      116      1      3.2      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0  
5   0     2       1  
6   3     1       0  
7   1     3       0  
8   0     3       0  
9   2     2       0

This Public Health Dataset represents people's medical records.

- age: age in years
- sex: (1 = male; 0 = female)
- cp: chest pain type
- trestbps: resting blood pressure (in mm Hg on admission to the hospital)
- chol: serum cholestoral in mg/dl
- fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
- restecg: resting electrocardiographic results
- thalach: maximum heart rate achieved
- exang: exercise induced angina (1 = yes; 0 = no)
- oldpeak: ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
- ca: number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
- target: 1 or 0. It refers to the presence of heart disease in the patient. It is integer valued 0 = no disease and 1 = disease.

The main aim is to build **a model that predicts a heart disease of a patient** (target column) based on independent variables.

## 3. Variables

Let's split the dataset into X and y, where 
- $X$ is a set of independent variables (all columns but the last one)
- $y$ is a dependent, or target variable (the last column)

In [6]:
# Filter out target column
X = df.iloc[:, :-1]

# Print X
X

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      52    1   0       125   212    0        1      168      0      1.0   
1      53    1   0       140   203    1        0      155      1      3.1   
2      70    1   0       145   174    0        1      125      1      2.6   
3      61    1   0       148   203    0        1      161      0      0.0   
4      62    0   0       138   294    1        1      106      0      1.9   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  
0         2   2     3  
1         0   0     3  
2         0   0     3  
3         2   1     3  
4         1   3     2  
...     ...  ..   ...  
1020      2   0     2  
1021      1   1     3  
1022      1   1     2  
1023      2   0     2  
1024      1   1     3  

[1025 rows x 13 columns]

In [7]:
# Select target column
y = df['target']

# Print y
y

0       0
1       0
2       0
3       0
4       0
       ..
1020    1
1021    0
1022    0
1023    1
1024    0
Name: target, Length: 1025, dtype: int64

## 4. Model

In this section we are going to 
- build an SVM model, 
- evaluate its accuracy, and 
- make a prediction

### 4.1. Building the model

Let's recall three simple steps:

- Split the $X$ & $y$ variables into train and test sets
- Initialize the model
- Train the model with the varialbes

In [90]:
# Split variables into train and test
X_train, X_test, y_train, y_test = train_test_split(X, #independent variables
                                                    y, #dependent variable
                                                    random_state = 3
                                                   )

In [91]:
# Initialize the model
clf = SVC(random_state=3, probability=True)

In [92]:
# Train the model
clf.fit(X_train, y_train)

SVC(probability=True, random_state=3)

### 4.2. Checking models accuracy

After the model has trained with the data, it's essential to understand how precisely it predicts heart disease. For that, we need to check model's accuracy. 

In [28]:
print(f'Accuracy of RBF-kernel SVC classifier on training set: {clf.score(X_train, y_train):.2f}')
print(f'Accuracy of RBF-kernel SVC classifier on test set: {clf.score(X_test, y_test):.2f}')

Accuracy of RBF-kernel SVC classifier on training set: 0.71
Accuracy of RBF-kernel SVC classifier on test set: 0.72


### 4.3. Making a prediction

Now that we know the model is accurate enough, we can predict whether or not a patient is having a heart disease by passing independent varialbes to the model. The method `predict` returns such a prediction - 0 for NO, 1 for YES.

In [29]:
clf.predict([[59, 1, 0, 101, 234, 0, 1, 143, 0, 3.4, 0, 0, 0]])

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/base.py:409: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


array([1])

In the array:
- value 0 means a patient does not have a heart disease, 
- value 1 means a patient does not have a heart disease.

We can also check the probability of a patient having a heart disease. The method `predict_proba` can be used to infer the class probabilities (i.e. the probability that a particular data point falls into the underlying classes).

NOTE: To be able to get probabilities, a hyperparameter `probability=True` should be specified during model initialization.

In [30]:
clf.predict_proba([[59, 1, 0, 101, 234, 0, 1, 143, 0, 3.4, 0, 0, 0]])

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/base.py:409: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


array([[0.4732713, 0.5267287]])

This array means that there is a 
- 47% chance of having NO heart disease, and 
- 53% chance of having a heart disease. 

Hence, the final prediction will be 1 (HAS heart disease).

## 5. Hyperparameters Tuning

Hyperparameter Tuning is one of the critical steps in model building. It helps in increasing model's accuracy even further.

### 5.1. Playing around with kernel hyperparameter

By default, SVC has `kernel=rbf`. Let's check `poly` kernel.

In [67]:
# Initialize the model
clf = SVC(kernel = 'poly',
          degree = 20)

In [68]:
# Train the model
clf.fit(X_train, y_train)

SVC(degree=20, kernel='poly')

In [94]:
# Check accuracy
print(f'Accuracy of poly-kernel SVC classifier on training set: {clf.score(X_train, y_train):.2f}')
print(f'Accuracy of poly-kernel SVC classifier on test set: {clf.score(X_train, y_train):.2f}')

Accuracy of poly-kernel SVC classifier on training set: 1.00
Accuracy of poly-kernel SVC classifier on test set: 1.00


In [97]:
# Check other kernels

for this_kernel in ['poly', 'linear', 'sigmoid']:
    
    # Initialize the model
    clf = SVC(kernel = this_kernel)
    
    # Train the model
    clf.fit(X_train, y_train)
    
    # Check accuracy
    print(f'Accuracy of {this_kernel}-kernel SVC classifier on training set: {clf.score(X_train, y_train):.2f}')
    print(f'Accuracy of {this_kernel}-kernel SVC classifier on test set: {clf.score(X_train, y_train):.2f}')

Accuracy of poly-kernel SVC classifier on training set: 0.71
Accuracy of poly-kernel SVC classifier on test set: 0.71
Accuracy of linear-kernel SVC classifier on training set: 0.85
Accuracy of linear-kernel SVC classifier on test set: 0.85
Accuracy of sigmoid-kernel SVC classifier on training set: 0.54
Accuracy of sigmoid-kernel SVC classifier on test set: 0.54


### 5.2. Playing around with gamma hyperparameter

In [93]:
for this_gamma in [0.001, 0.01, 0.1]:
    
    #initialize a model
    clf = SVC(gamma=this_gamma)
    
    #fit the model
    clf.fit(X_train, y_train)
    
    #print results
    print(f'SVM with gamma = {this_gamma}')
    print(f'Accuracy of SVM (RBF) classifier on training set: {clf.score(X_train, y_train):.2f}')
    print(f'Accuracy of SVM (RBF) classifier on test set: {clf.score(X_test, y_test):.2f}\n')

SVM with gamma = 0.001
Accuracy of SVM (RBF) classifier on training set: 0.81
Accuracy of SVM (RBF) classifier on test set: 0.79

SVM with gamma = 0.01
Accuracy of SVM (RBF) classifier on training set: 0.99
Accuracy of SVM (RBF) classifier on test set: 0.98

SVM with gamma = 0.1
Accuracy of SVM (RBF) classifier on training set: 1.00
Accuracy of SVM (RBF) classifier on test set: 1.00



### 5.3. Playing around with both C and gamma

In [88]:
for this_gamma in [0.001, 0.01, 0.1]:
    
    for this_C in [0.1, 1, 15]:
        
        #initialize a model
        clf = SVC(gamma=this_gamma, C = this_C)
        
        #fit the model
        clf.fit(X_train, y_train)

        #print results
        print(f'SVM with gamma = {this_gamma} & C = {this_C}')
        print(f'Accuracy of SVM (RBF) classifier on training set: {clf.score(X_train, y_train):.2f}')
        print(f'Accuracy of SVM (RBF) classifier on test set: {clf.score(X_test, y_test):.2f}\n')

SVM with gamma = 0.001 & C = 0.1
Accuracy of SVM (RBF) classifier on training set: 0.71
Accuracy of SVM (RBF) classifier on test set: 0.70

SVM with gamma = 0.001 & C = 1
Accuracy of SVM (RBF) classifier on training set: 0.81
Accuracy of SVM (RBF) classifier on test set: 0.79

SVM with gamma = 0.001 & C = 15
Accuracy of SVM (RBF) classifier on training set: 0.96
Accuracy of SVM (RBF) classifier on test set: 0.92

SVM with gamma = 0.01 & C = 0.1
Accuracy of SVM (RBF) classifier on training set: 0.56
Accuracy of SVM (RBF) classifier on test set: 0.54

SVM with gamma = 0.01 & C = 1
Accuracy of SVM (RBF) classifier on training set: 0.99
Accuracy of SVM (RBF) classifier on test set: 0.98

SVM with gamma = 0.01 & C = 15
Accuracy of SVM (RBF) classifier on training set: 1.00
Accuracy of SVM (RBF) classifier on test set: 1.00

SVM with gamma = 0.1 & C = 0.1
Accuracy of SVM (RBF) classifier on training set: 0.51
Accuracy of SVM (RBF) classifier on test set: 0.51

SVM with gamma = 0.1 & C = 1
Ac

### 5.4. All Hyperparameters of SVM

There are many hyperparameters in SVC. They are all presented below  - as you can see, they have values by default. Feel free to play around with them. To learn more about hyperparameters and their meaning, visit https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.

In [89]:
clf = SVC(C=1.0, 
          kernel='rbf', 
          degree=3, 
          gamma='scale', 
          coef0=0.0, 
          shrinking=True, 
          probability=False, # We set to true to be able to use predict_prob
          tol=0.001, 
          cache_size=200, 
          class_weight=None, 
          verbose=False, 
          max_iter=-1, 
          decision_function_shape='ovr', 
          break_ties=False, 
          random_state=None)

### 6. SVMs with normalized data

(feature preprocessing using minmax scaling)

In [87]:
#initialize scaler
scaler = MinMaxScaler()

#transform variables
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

#initialize the model
clf = SVC()

#train the model
clf.fit(X_train_scaled, y_train)

#print the results
print('Accuracy of SVM (with MinMax scaling) classifier on training set: {:.2f}'
     .format(clf.score(X_train_scaled, y_train)))
print('Accuracy of SVM (with MinMax scaling) classifier on test set: {:.2f}'
     .format(clf.score(X_test_scaled, y_test)))

Accuracy of SVM (with MinMax scaling) classifier on training set: 0.91
Accuracy of SVM (with MinMax scaling) classifier on test set: 0.88
